File: ai-layer/psychology/feedback_loop.py
"""Online feedback loop (L7). Stdlib EMA per-intent confidence, mirroring the Java FeedbackLoop.

The production version uses River for true online learning; the contract (record / weight) is stable.
"""
from __future__ import annotations

from typing import Dict

ALPHA = 0.2
START = 0.5


class FeedbackLoop:
    def __init__(self) -> None:
        self._weights: Dict[str, float] = {}

    def record(self, intent: str, positive: bool) -> None:
        if not intent:
            return
        target = 1.0 if positive else 0.0
        current = self._weights.get(intent, START)
        self._weights[intent] = ALPHA * target + (1 - ALPHA) * current

    def weight(self, intent: str) -> float:
        return self._weights.get(intent, START)

    def snapshot(self) -> Dict[str, float]:
        return dict(self._weights)

File: backend/src/main/java/com/ultron/intelligence/psychology/FeedbackLoop.java
package com.ultron.intelligence.psychology;

import com.ultron.intelligence.psychology.IntentClassifier.Intent;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import org.springframework.stereotype.Component;

/**
 * Online feedback loop (L7 — The Mind). Records thumbs-up / thumbs-down on the agent's handling of
 * each {@link Intent} and maintains a running confidence weight per intent. Phase 1 is a simple
 * in-memory exponential update (the offline analogue of the Python River online-learning model);
 * it lets downstream components prefer intents the owner has reinforced.
 */
@Component
public class FeedbackLoop {

    /** EMA smoothing factor for each new signal. */
    private static final double ALPHA = 0.2;
    private static final double START = 0.5;

    private final Map<Intent, Double> weights = new ConcurrentHashMap<>();

    /** Record feedback for an intent: {@code positive=true} reinforces, {@code false} penalises. */
    public void record(Intent intent, boolean positive) {
        if (intent == null) {
            return;
        }
        double target = positive ? 1.0 : 0.0;
        weights.merge(intent, ALPHA * target + (1 - ALPHA) * START,
            (oldVal, ignored) -> ALPHA * target + (1 - ALPHA) * oldVal);
    }

    /** Current confidence weight in [0,1] for an intent (defaults to the neutral start). */
    public double weight(Intent intent) {
        return weights.getOrDefault(intent, START);
    }

    /** Snapshot of all learned weights. */
    public Map<Intent, Double> snapshot() {
        return Map.copyOf(weights);
    }
}
